File: benchmarkDS/ourMethod/OPENAI_gpt-5-2025-08-07/Lse94.java
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

public class Lse94 {

    // Security parameters
    private static final int PBKDF2_ITERATIONS = 210_000;
    private static final int SALT_LENGTH_BYTES = 16;
    private static final int DERIVED_KEY_LENGTH_BYTES = 32;

    private static final Pattern USERNAME_PATTERN = Pattern.compile("^[A-Za-z0-9_]{3,32}$");

    public static final class UserRecord {
        public final String username;
        public final String saltB64;
        public final String hashB64;
        public final int iterations;

        public UserRecord(String username, String saltB64, String hashB64, int iterations) {
            this.username = username;
            this.saltB64 = saltB64;
            this.hashB64 = hashB64;
            this.iterations = iterations;
        }
    }

    // In-memory database simulation
    public static final class InMemoryDatabase {
        private final Map<String, UserRecord> users = Collections.synchronizedMap(new HashMap<>());

        public boolean insert(UserRecord rec) {
            Objects.requireNonNull(rec, "record");
            synchronized (users) {
                if (users.containsKey(rec.username)) {
                    return false;
                }
                users.put(rec.username, rec);
                return true;
            }
        }

        public UserRecord get(String username) {
            synchronized (users) {
                return users.get(username);
            }
        }

        public int size() {
            synchronized (users) {
                return users.size();
            }
        }
    }

    // Public API: register user (username + password). Returns stored record on success, else null.
    public static UserRecord registerUser(InMemoryDatabase db, String username, char[] password) {
        if (db == null) {
            return null;
        }
        if (!isValidUsername(username)) {
            wipe(password);
            return null;
        }
        if (!isStrongPassword(password)) {
            wipe(password);
            return null;
        }

        byte[] salt = null;
        byte[] dk = null;
        try {
            salt = generateSalt(SALT_LENGTH_BYTES);
            dk = hashPassword(password, salt, PBKDF2_ITERATIONS, DERIVED_KEY_LENGTH_BYTES);

            String saltB64 = Base64.getEncoder().encodeToString(salt);
            String hashB64 = Base64.getEncoder().encodeToString(dk);

            UserRecord rec = new UserRecord(username, saltB64, hashB64, PBKDF2_ITERATIONS);
            boolean ok = db.insert(rec);
            if (!ok) {
                return null;
            }
            return rec;
        } catch (GeneralSecurityException e) {
            return null;
        } finally {
            wipe(password);
            wipe(salt);
            wipe(dk);
        }
    }

    // Helpers

    private static boolean isValidUsername(String username) {
        if (username == null) return false;
        if (username.length() < 3 || username.length() > 32) return false;
        return USERNAME_PATTERN.matcher(username).matches();
    }

    private static boolean isStrongPassword(char[] password) {
        if (password == null || password.length < 12) return false;
        boolean hasLower = false, hasUpper = false, hasDigit = false, hasSpecial = false, hasSpace = false;
        for (char c : password) {
            if (Character.isLowerCase(c)) hasLower = true;
            else if (Character.isUpperCase(c)) hasUpper = true;
            else if (Character.isDigit(c)) hasDigit = true;
            else if (Character.isWhitespace(c)) hasSpace = true;
            else hasSpecial = true;
        }
        if (hasSpace) return false;
        return hasLower && hasUpper && hasDigit && hasSpecial;
    }

    private static byte[] generateSalt(int len) {
        byte[] salt = new byte[len];
        new SecureRandom().nextBytes(salt);
        return salt;
    }

    private static byte[] hashPassword(char[] password, byte[] salt, int iterations, int dkLen)
            throws GeneralSecurityException {
        PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, dkLen * 8);
        try {
            SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            return skf.generateSecret(spec).getEncoded();
        } finally {
            spec.clearPassword();
        }
    }

    private static void wipe(byte[] data) {
        if (data != null) {
            for (int i = 0; i < data.length; i++) data[i] = 0;
        }
    }

    private static void wipe(char[] data) {
        if (data != null) {
            for (int i = 0; i < data.length; i++) data[i] = 0;
        }
    }

    public static void main(String[] args) {
        InMemoryDatabase db = new InMemoryDatabase();

        // 5 test cases
        String[] usernames = {
            "alice_01",
            "bobUser_22",
            "x",             // invalid username (too short)
            "charlie_33",
            "alice_01"       // duplicate username
        };
        String[] passwords = {
            "Sup3rStr0ng!Pwd",     // strong
            "An0ther$trongPass!",  // strong
            "Short!1A",            // weak (too short)
            "NoSpacesAllowed!9",   // strong
            "Sup3rStr0ng!Pwd"      // strong but duplicate username
        };

        for (int i = 0; i < 5; i++) {
            char[] pwd = passwords[i].toCharArray();
            UserRecord rec = registerUser(db, usernames[i], pwd);
            System.out.println("Test " + (i + 1) + " - username=" + usernames[i] + " -> " + (rec != null ? "stored" : "rejected"));
        }

        // Show how many users stored
        System.out.println("Total stored users: " + db.size());

        // Optionally show a stored record without secrets (no plaintext, only metadata)
        UserRecord alice = db.get("alice_01");
        if (alice != null) {
            System.out.println("Stored record for alice_01: iterations=" + alice.iterations +
                    ", salt(len)=" + (alice.saltB64 != null ? alice.saltB64.length() : 0) +
                    ", hash(len)=" + (alice.hashB64 != null ? alice.hashB64.length() : 0));
        }
    }
}

File: benchmarkDS/ourMethod/OPENAI_gpt-5-2025-08-07/Lse94.py
import base64
import hashlib
import hmac
import re
import secrets
from typing import Optional, Dict, Any


PBKDF2_ITERATIONS = 210_000
SALT_LEN = 16
DK_LEN = 32
USERNAME_RE = re.compile(r"^[A-Za-z0-9_]{3,32}$")


class InMemoryDatabase:
    def __init__(self) -> None:
        self._users: Dict[str, Dict[str, Any]] = {}

    def insert(self, record: Dict[str, Any]) -> bool:
        username = record.get("username")
        if username in self._users:
            return False
        self._users[username] = record
        return True

    def get(self, username: str) -> Optional[Dict[str, Any]]:
        return self._users.get(username)

    def size(self) -> int:
        return len(self._users)


def is_valid_username(username: str) -> bool:
    if not isinstance(username, str):
        return False
    if not 3 <= len(username) <= 32:
        return False
    return bool(USERNAME_RE.fullmatch(username))


def is_strong_password(password: str) -> bool:
    if not isinstance(password, str):
        return False
    if len(password) < 12:
        return False
    has_lower = any(c.islower() for c in password)
    has_upper = any(c.isupper() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_space = any(c.isspace() for c in password)
    has_special = any(not c.isalnum() and not c.isspace() for c in password)
    if has_space:
        return False
    return has_lower and has_upper and has_digit and has_special


def hash_password_with_new_salt(password: str) -> Dict[str, Any]:
    salt = secrets.token_bytes(SALT_LEN)
    dk = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, PBKDF2_ITERATIONS, dklen=DK_LEN)
    record = {
        "salt_b64": base64.b64encode(salt).decode("ascii"),
        "hash_b64": base64.b64encode(dk).decode("ascii"),
        "iterations": PBKDF2_ITERATIONS,
    }
    # zeroize local variables where possible
    # note: Python strings/bytes are immutable; we avoid keeping them around
    return record


def register_user(db: InMemoryDatabase, username: str, password: str) -> Optional[Dict[str, Any]]:
    if db is None:
        return None
    if not is_valid_username(username):
        return None
    if not is_strong_password(password):
        return None

    record_parts = hash_password_with_new_salt(password)
    record = {
        "username": username,
        "salt_b64": record_parts["salt_b64"],
        "hash_b64": record_parts["hash_b64"],
        "iterations": record_parts["iterations"],
    }
    if not db.insert(record):
        return None
    return record


def _ct_equal(a: bytes, b: bytes) -> bool:
    return hmac.compare_digest(a, b)


def main():
    db = InMemoryDatabase()

    tests = [
        ("alice_01", "Sup3rStr0ng!Pwd"),
        ("bobUser_22", "An0ther$trongPass!"),
        ("x", "Short!1A"),
        ("charlie_33", "NoSpacesAllowed!9"),
        ("alice_01", "Sup3rStr0ng!Pwd"),
    ]

    for i, (u, p) in enumerate(tests, 1):
        rec = register_user(db, u, p)
        print(f"Test {i} - username={u} -> {'stored' if rec else 'rejected'}")

    print(f"Total stored users: {db.size()}")
    alice = db.get("alice_01")
    if alice:
        print(
            "Stored record for alice_01: "
            f"iterations={alice['iterations']}, salt(len)={len(alice['salt_b64'])}, hash(len)={len(alice['hash_b64'])}"
        )


if __name__ == "__main__":
    main()